transport/src/main/java/io/netty/channel/AbstractChannel.java

```diff
@@ -1048,4 +1065,54 @@ boolean setClosed() {
             return super.trySuccess();
         }
     }
+
+    private static final class AutoFlushTask implements Runnable {
+
+        private final AbstractChannel channel;
+        private final SingleThreadEventLoop eventLoop;
+        private boolean writePending;
+        private boolean running;
+
+        AutoFlushTask(AbstractChannel channel) {
+            this.channel = channel;
+            assert channel.eventLoop() instanceof SingleThreadEventLoop;
+            eventLoop = (SingleThreadEventLoop) channel.eventLoop();
+        }
+
+        @Override
+        public void run() {
+            running = true;
+            try {
+                if (writePending) {
+                    writePending = false;
+                    channel.flush();
+                    if (writePending) {
+                        // There may be cases where a handler writes more data from within the flush() method and then
+                        // does not flush. In such a case, if we call flush() on the pipeline again, then it may end up
+                        // in an infinite loop. In order to prevent such an issue, we directly flush the unsafe.
+                        channel.unsafe.flush();
+                        writePending = false;
+                    }
+
+                    channel.pipeline.postAutoFlush();
```

Comment:
Auto flush will not really be broken, as the `postAutoFlush()` handles eventloop wakeup and not `AutoFlush` tasks scheduling. Having said that, I agree that it is better to put it in `finally` block.